core/src/pretty.rs

```diff
@@ -884,63 +966,601 @@ where
+                  4
+                ]"
+            },
+        );
+    }
+
+    #[test]
+    fn pretty_match() {
+        assert_long_short_term(
+            "match { 'A => a, 'B => b, 'C => c, }",
+            indoc! {"
+                match {
+                  'A =>
+                    a,
+                  'B =>
+                    b,
+                  'C =>
+                    c,
+                }"
+            },
+        );
+    }
+
+    #[test]
+    fn pretty_record() {
+        assert_long_short_term("{}", "{}");
+        assert_long_short_term(
+            "{ a = b, c = d, }",
+            indoc! {"
+                {
+                  a =
+                    b,
+                  c =
+                    d,
+                }"
+            },
+        );
+        assert_long_short_term(
+            r#"{ a | String | force = b, c | Number | doc "" = d, }"#,
+            indoc! {r#"
+                {
+                  a
+                    | String
+                    | force
+                    =
+                    b,
+                  c
+                    | Number
+                    | doc ""
+                    =
+                    d,
+                }"#
+            },
+        );
+        assert_long_short_term(
+            "{ a = b, .. }",
+            indoc! {"
+                {
+                  a =
+                    b,
+                  ..
+                }"
+            },
+        );
+        assert_long_short_term(
+            r#"{ a = b, "%{a}" = c, .. }"#,
+            indoc! {r#"
+                {
+                  a =
+                    b,
+                  "%{a}" =
+                    c,
+                  ..
+                }"#
+            },
+        );
+        assert_long_short_term(
+            r#"{ "=" = a, }"#,
+            indoc! {r#"
+                {
+                  "=" =
+                    a,
+                }"#
+            },
+        );
+    }
+
+    #[test]
+    fn pretty_let() {
+        assert_long_short_term(
+            "let rec foo | String = c in {}",
+            indoc! {"
+                let rec foo
+                  | String
+                  = c
+                  in
+                {}"
+            },
+        );
+        assert_long_short_term(
+            "let foo = c bar in {}",
+            indoc! {"
+                let foo
+                  = c
+                    bar
+                  in
+                {}"
+            },
+        );
+        assert_long_short_term(
+            "let foo | String = c bar in {}",
+            indoc! {"
+                let foo
+                  | String
+                  = c
+                    bar
+                  in
+                {}"
+            },
+        );
+    }
+
+    #[test]
+    fn pretty_multiline_strings() {
+        let t: RichTerm = Term::StrChunks(vec![StrChunk::Literal("\n1.".to_owned())]).into();
+        assert_eq!(format!("{t}"), "m%\"\n  \n  1.\n\"%");
```

Comment:
where are these extra newlines coming from?

Reply:
When `\n1.` gets formatted as a multiline string, the result is
```
m%"

  1.
"%
```

The first newline after `m%"` is removed by the parser, as is the last newline before `"%`. I tried for a while to come up with a good heuristic to decide if it would be better to keep `"\n1."` as a single-line string with escapes, but I couldn't find anything reasonable to distinguish it from for example doc-strings that should be printed using the `m%""%` syntax...

Unfortunately, I couldn't use `indoc!` here because `pretty.rs` insists on putting two spaces after every indented newline, even if the line is otherwise empty. But `indoc!` would remove those spaces.